Sparse CSR: Add backward for torch.sparse.sampled_addmm

IvanYashchuk opened the pull request:
This PR adds support for backward for `torch.sparse.sampled_addmm`. Gradcheck doesn't work with sparse CSR inputs so we're testing here directly the `.backward()` call against a dense reference computation.


Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76591
* #76590
* #76589
* __->__ #68084

IvanYashchuk committed:
Sparse CSR: Add backward for torch.sparse.sampled_addmm

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"




[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -460,9 +460,9 @@ Tensor empty_like_sparse_csr(
         self.col_indices().clone(),
         at::empty(self.values().sizes(), options.layout(kStrided)),
```
I think we need to pass "device" to empty as well? Neither `_sparse_csr_tensor_unsafe` nor `SparseCsrTensorImpl::set_member_tensors` do a conversion to the correct device.

cpuhrsch replied:
No, I just fully missed what "layout" actually does here....

cpuhrsch commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2491,6 +2491,25 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad, **kwargs):
+        for index_dtype in (torch.int32, torch.int64):
+            for m, n, k in itertools.product([0, 5], repeat=3):
+                yield SampleInput(
+                        make_tensor((m, n), device=device, dtype=dtype).to_sparse_csr().requires_grad_(requires_grad),
```
you could make sure that there are some zeros in here by e.g. calling relu() on the result. it might make for a more powerful test, since I think make_tensor will just give you randn. We can do this in a follow-up.

IvanYashchuk replied:
Sure, I will change that.

cpuhrsch committed:
Lint fixes

cpuhrsch committed:
Lint fixes

cpuhrsch commented:
Failure looks real, I think it's because we don't have gradcheck support yet.

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"

This PR adds support for backward for `torch.sparse.sampled_addmm`. Gradcheck doesn't work with sparse CSR inputs so we're testing here directly the `.backward()` call against a dense reference computation.





[ghstack-poisoned]

IvanYashchuk commented:
> [Failure](https://github.com/pytorch/pytorch/runs/6219335743?check_suite_focus=true) looks real, I think it's because we don't have gradcheck support yet. Alternatively you can disable [this check](https://github.com/pytorch/pytorch/blob/4048d4cdd2a579cf16bc86bc7e76f9ab40d80226/test/test_ops.py#L181-L184) for this particular OpInfo, but you'd probably need another flag, so it might be more elegant to add gradcheck support instead.

This particular failure is fixed simply by setting `supports_autograd=True`, I overlooked this initially. That's right for now many tests are skipped and they can be enabled once gradcheck for sparse CSR inputs is implemented. But we don't need gradcheck right now to test correctness since we can compare against the result computed with dense strided inputs.

IvanYashchuk committed:
Update on "Sparse CSR: Add backward for torch.sparse.sampled_addmm"

This PR adds support for backward for `torch.sparse.sampled_addmm`. Gradcheck doesn't work with sparse CSR inputs so we're testing here directly the `.backward()` call against a dense reference computation.





[ghstack-poisoned]

IvanYashchuk commented:
@pytorchbot merge this

cpuhrsch commented:
@pytorchbot merge this

malfet commented:
Going to propose a dummy change in order to trigger lint  (as workflow dispatch is not yet enabled for lint jobs)

malfet commented:
@pytorchbot force merge this (as lint were never scheduled)
